conceptnet_sme/semantic_matching.py

```diff
@@ -966,40 +1273,45 @@ def model_copier(model, device):
         sampler=CyclingSampler(dataset),
         pin_memory=True,
     )
+    if validation_dataset is not None:
+        validation_data_loader = DataLoader(
+            validation_dataset,
+            batch_size=BATCH_SIZE,
+            drop_last=False,
+            num_workers=num_batch_workers,
+            collate_fn=validation_dataset.collate_batch,
+            sampler=CyclingSampler(validation_dataset),
+            pin_memory=True
+        )
+        # Since we want to step through the (infinite, cyclic) sequence of
+        # validation batches generated by this data loader, and take fixed-
+        # size slices from that sequence, we convert the data loader (which
+        # is iterable) into an iterator.  Note that (as the sequence is
+        # infinite) this iterator will never raise StopIteration.
+        validation_data_iterator = iter(validation_data_loader)
```

Comment:
Is this being too clever with the way we iterate over the data? It seems to me that we'd want the validation set to be a fixed, finite set.

Reply:
I did it this way because not every edge from a validation file will necessarily survive the filters (e.g. the relation must be in COMMON_RELATIONS, both the terms must be in the embedding) we impose, so it's easiest to use a large validation file to be sure some edges are left.  But then using the whole validation set is slow.  If you don't find that convincing let me know, and I'll make the change you propose.

Reply:
I'd prefer to be more certain about what's in the validation set, even if we evaluate on a sub-sample of it at any given moment. For example, we could do the filtering first and save the result.

Reply:
OK, done.